postgresql/postgresql_plugin.cc

```diff
@@ -0,0 +1,832 @@
+            POSTGRESQL_VERBOSE(cerr << tableDescription << endl;)
+            res = PQexec(conn, tableDescription.c_str());
+            if (PQresultStatus(res) != PGRES_COMMAND_OK) {
+                string errorMsg(PQresultErrorMessage(res));
+                PQclear(res);
+                PQfinish(conn);
+                throw HttpReturnException(400, "Could not create PostgreSQL table:  ", errorMsg);
+            }
+
+            POSTGRESQL_VERBOSE(cerr << "table " << config_.tableName << " created!" << endl;)
+            PQclear(res);
+            PQfinish(conn);
+        }
+    }
+    
+    virtual ~PostgresqlRecorderDataset()
+    {
+    }
+
+    virtual Any getStatus() const override
+    {
+        return string("ok");
+    }
+
+    string getPostgresTypeString(CellValue v)
+    {
+        switch (v.cellType())
+        {
+            case CellValue::EMPTY:
+            case CellValue::INTEGER:
+                return "integer";
+            case CellValue::FLOAT:
+                return "float8";
+            case CellValue::ASCII_STRING:
+            case CellValue::UTF8_STRING:
+            case CellValue::PATH:
+                return "varchar";
+            case CellValue::TIMESTAMP:
+                return "timestamptz";
+            case CellValue::TIMEINTERVAL:
+                return "interval";
+            case CellValue::BLOB:
+                return "varbit";
+            default:
+                throw HttpReturnException(500, "Unknown cell type in getPostgresTypeString");
+        }
+    }
+
+    void alterColumns(pg_conn* conn, const std::vector<std::tuple<ColumnName, CellValue, Date> > & vals)
+    {
+        for (auto& p : vals) {
+            ColumnName column = std::get<0>(p);
+            if (insertedColumns.count(column) == 0){
+                insertedColumns.insert(column);
+                string alterString = "ALTER TABLE " +
+                                     config_.tableName +
+                                     " ADD COLUMN " +
+                                     column.toUtf8String().rawString() + 
+                                     " "
+                                     + getPostgresTypeString(std::get<1>(p));
+
+                POSTGRESQL_VERBOSE(cerr << alterString << endl;)
+
+                auto res = PQexec(conn, alterString.c_str());
+                if (PQresultStatus(res) != PGRES_COMMAND_OK) {
+                    string errorMsg(PQresultErrorMessage(res));
+                    PQclear(res);
+                    PQfinish(conn);
+                    throw HttpReturnException(400, "Could not alter PostgreSQL table:  ", errorMsg);
+                }
+
+                PQclear(res);
+            }
+        }
+    }
+
+    void insertRowPostgresql(pg_conn* conn, const std::vector<std::tuple<ColumnName, CellValue, Date> > & vals) 
+    {
+        alterColumns(conn, vals);
+
+        string insertString = "INSERT INTO " + config_.tableName + " (";
```

Comment:
There is a lot of string manipulation on each insertion, I wonder if we should not have a preallocated buffer to avoid reallocation.  May be in a second pass.